src/main/scala/mesosphere/marathon/core/election/ElectionService.scala

```diff
@@ -91,10 +105,155 @@ trait ElectionCandidate {
   def startLeadership(): Unit
 }
 
-/** Local leadership events. They are not delivered via the event endpoints. */
-sealed trait LocalLeadershipEvent
+class ElectionServiceImpl(
+    eventStream: EventStream,
+    hostPort: String,
+    leaderEventsSource: Source[LeadershipState, Cancellable],
+    crashStrategy: CrashStrategy,
+    electionEC: ExecutionContext
+)(implicit system: ActorSystem) extends ElectionService with StrictLogging {
+
+  import ElectionService._
+  @volatile private[this] var lastState: LeadershipState = LeadershipState.Standby(None)
+  implicit private lazy val materializer = ActorMaterializer()
+  var leaderSubscription: Option[Cancellable] = None
+
+  def subscribe(subscriber: ActorRef): Unit = {
+    eventStream.subscribe(subscriber, classOf[LeadershipTransition])
+    val currentState = if (isLeader) LeadershipTransition.ElectedAsLeader else LeadershipTransition.Standby
+    subscriber ! currentState
+  }
+
+  def unsubscribe(subscriber: ActorRef): Unit = {
+    eventStream.unsubscribe(subscriber, classOf[LeadershipTransition])
+  }
+
+  override def isLeader: Boolean =
+    lastState == LeadershipState.ElectedAsLeader
 
-object LocalLeadershipEvent {
-  case object ElectedAsLeader extends LocalLeadershipEvent
-  case object Standby extends LocalLeadershipEvent
+  override def localHostPort: String = hostPort
+
+  override def leaderHostPort: Option[String] = lastState match {
+    case LeadershipState.ElectedAsLeader =>
+      Some(hostPort)
+    case LeadershipState.Standby(currentLeader) =>
+      currentLeader
+  }
+
+  /**
+    * Releases leadership
+    *
+    * Has no effect if called before offerLeadership
+    */
+  override def abdicateLeadership(): Unit = {
+    leaderSubscription.foreach(_.cancel())
+  }
+
+  private def initializeStream(leadershipTransitionsFlow: Flow[LeadershipState, LeadershipTransition, NotUsed]) = {
+    val localEventListenerSink = Sink.foreach[LeadershipState] { event =>
+      lastState = event
+    }
+    val localTransitionSink = Sink.foreach[LeadershipTransition] { e =>
+      eventStream.publish(e)
+      if (e == LeadershipTransition.Standby) {
```

Comment:
Is this true? I thought it should be `if(lastState == LeadershipTransition.ElectedAsLeader && e == LeadershipTransition.Standby)`.

Reply:
`LeadershipTransition.Standby` is only published, by definition, if leadership was previously had.

Enriched documentation for these events to make more clear.

There is already a test for this.

✅

Reply:
Thanks for clearing this up. I did not know that the graph would ensure this.